File: session/opencv/p4/main.cpp
#include <opencv2/opencv.hpp>
#include <iostream>

int main() {
    // Cargar imagen
    cv::Mat img = cv::imread("hand_dataset-5.png");
    if (img.empty()) {
        std::cerr << "No se pudo cargar la imagen" << std::endl;
        return -1;
    }

    // Convertir a escala de grises
    cv::Mat gray;
    cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);

    // Umbralización usando Otsu
    cv::Mat mask;
    cv::threshold(gray, mask, 0, 255, cv::THRESH_BINARY | cv::THRESH_OTSU);
    cv::bitwise_not(mask, mask);  // Invertir máscara

    // Detección de contornos
    std::vector<std::vector<cv::Point>> contours;
    std::vector<cv::Vec4i> hierarchy;
    cv::findContours(mask, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

    // Dibujar contornos sobre la imagen original
    cv::Mat img_contours = img.clone();
    cv::drawContours(img_contours, contours, -1, cv::Scalar(0, 255, 0), 2);

    cv::Mat img_defects = img.clone();
    for (size_t i = 0; i < contours.size(); ++i) {
        // Eliminar contornos con área < 500
        if (cv::contourArea(contours[i]) < 500) continue;

        // Calcular el convex hull
        std::vector<cv::Point> hull;
        cv::convexHull(contours[i], hull, true);

        // Dibujar convex hull en azul
        cv::polylines(img_defects, hull, true, cv::Scalar(255, 0, 0), 2);

        // Calcular defectos de convexidad
        std::vector<int> hull_indices;
        cv::convexHull(contours[i], hull_indices, false);
        if (hull_indices.size() > 3) {
            cv::Mat defects;
            cv::convexityDefects(contours[i], hull_indices, defects);

            // Dibujar defectos con profundidad >= 10
            if (!defects.empty()) {
                for (int j = 0; j < defects.rows; ++j) {
                    int start_idx = defects.at<int>(j, 0);
                    int end_idx = defects.at<int>(j, 1);
                    int far_idx = defects.at<int>(j, 2);
                    float depth = defects.at<int>(j, 3) / 256.0f;

                    if (depth < 10) continue;

                    cv::Point start = contours[i][start_idx];
                    cv::Point end = contours[i][end_idx];
                    cv::Point far = contours[i][far_idx];

                    // Dibujar triángulo del defecto
                    cv::line(img_defects, start, far, cv::Scalar(0, 255, 0), 1);
                    cv::line(img_defects, far, end, cv::Scalar(0, 255, 0), 1);
                    cv::line(img_defects, start, end, cv::Scalar(0, 255, 255), 1);

                    // Mostrar profundidad
                    cv::putText(img_defects, cv::format("%.1f", depth),
                                cv::Point(far.x, far.y + 15),
                                cv::FONT_HERSHEY_SIMPLEX, 0.4,
                                cv::Scalar(255, 255, 255), 1, cv::LINE_AA);
                }
            }
        }
    }

    // === 6. Mostrar resultados ===
    cv::imshow("Original", img);
    cv::imshow("Umbral", mask);
    cv::imshow("Contornos", img_contours);
    cv::imshow("Defectos", img_defects);

    cv::waitKey(0);
    cv::destroyAllWindows();

    return 0;
}

File: session/opencv/p4/main.py
import cv2
import numpy as np

# --- 1. Cargar imagen ---
img = cv2.imread("hand_dataset-5.png")
if img is None:
    print("No se pudo cargar la imagen")
    exit(-1)

# --- 2. Convertir a escala de grises ---
gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# --- 3. Umbralización usando Otsu ---
_, mask = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
mask = cv2.bitwise_not(mask)  # Invertir máscara

# --- 4. Detección de contornos ---
contours, hierarchy = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

# --- 5. Dibujar contornos sobre la imagen original ---
img_contours = img.copy()
cv2.drawContours(img_contours, contours, -1, (0, 255, 0), 2)

img_defects = img.copy()
for cnt in contours:
    # Eliminar contornos con área < 500
    if cv2.contourArea(cnt) < 500:
        continue

    # Calcular convex hull
    hull = cv2.convexHull(cnt, returnPoints=True)
    cv2.polylines(img_defects, [hull], True, (255, 0, 0), 2)

    # Calcular defectos de convexidad
    hull_indices = cv2.convexHull(cnt, returnPoints=False)
    if len(hull_indices) > 3:
        defects = cv2.convexityDefects(cnt, hull_indices)
        if defects is not None:
            for i in range(defects.shape[0]):
                s, e, f, d = defects[i, 0]
                depth = d / 256.0
                if depth < 10:
                    continue

                start = tuple(cnt[s][0])
                end = tuple(cnt[e][0])
                far = tuple(cnt[f][0])

                # Dibujar triángulo del defecto
                cv2.line(img_defects, start, far, (0, 255, 0), 1)
                cv2.line(img_defects, far, end, (0, 255, 0), 1)
                cv2.line(img_defects, start, end, (0, 255, 255), 1)

                # Mostrar profundidad
                text_pos = (far[0], far[1] + 15)
                cv2.putText(img_defects, f"{depth:.1f}", text_pos,
                            cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255, 255, 255), 1, cv2.LINE_AA)

# --- 6. Mostrar resultados ---
cv2.imshow("Original", img)
cv2.imshow("Umbral", mask)
cv2.imshow("Contornos", img_contours)
cv2.imshow("Defectos", img_defects)

cv2.waitKey(0)
cv2.destroyAllWindows()
